File: frontend/appflowy_flutter/lib/plugins/database_view/application/group/group_service.dart
import 'package:appflowy_backend/dispatch/dispatch.dart';
import 'package:appflowy_backend/protobuf/flowy-database2/group.pb.dart';
import 'package:appflowy_backend/protobuf/flowy-error/errors.pb.dart';
import 'package:dartz/dartz.dart';

class GroupBackendService {
  final String viewId;

  GroupBackendService(this.viewId);

  Future<Either<Unit, FlowyError>> groupByField({
    required String fieldId,
  }) {
    final payload = GroupByFieldPayloadPB.create()
      ..viewId = viewId
      ..fieldId = fieldId;

    return DatabaseEventSetGroupByField(payload).send();
  }

  Future<Either<Unit, FlowyError>> updateGroup({
    required String groupId,
    required String fieldId,
    String? name,
    bool? visible,
  }) {
    final payload = UpdateGroupPB.create()
      ..fieldId = fieldId
      ..viewId = viewId
      ..groupId = groupId;

    if (name != null) {
      payload.name = name;
    }
    if (visible != null) {
      payload.visible = visible;
    }
    return DatabaseEventUpdateGroup(payload).send();
  }

  Future<Either<Unit, FlowyError>> createGroup({
    required String name,
    String groupConfigId = "",
  }) {
    final payload = CreateGroupPayloadPB.create()
      ..viewId = viewId
      ..name = name;

    return DatabaseEventCreateGroup(payload).send();
  }

  Future<Either<Unit, FlowyError>> deleteGroup({
    required String groupId,
  }) {
    final payload = DeleteGroupPayloadPB.create()
      ..viewId = viewId
      ..groupId = groupId;

    return DatabaseEventDeleteGroup(payload).send();
  }
}

File: frontend/appflowy_tauri/src/appflowy_app/components/database/application/group/group_service.ts
import {
  DatabaseViewIdPB,
  GroupByFieldPayloadPB,
  MoveGroupPayloadPB,
  UpdateGroupPB,
} from '@/services/backend';
import {
  DatabaseEventGetGroups,
  DatabaseEventMoveGroup,
  DatabaseEventSetGroupByField,
  DatabaseEventUpdateGroup,
} from '@/services/backend/events/flowy-database2';
import { Group, pbToGroup } from './group_types';

export async function getGroups(viewId: string): Promise<Group[]> {
  const payload = DatabaseViewIdPB.fromObject({ value: viewId });

  const result = await DatabaseEventGetGroups(payload);

  return result.map(value => value.items.map(pbToGroup)).unwrap();
}

export async function setGroupByField(viewId: string, fieldId: string): Promise<void> {
  const payload = GroupByFieldPayloadPB.fromObject({
    view_id: viewId,
    field_id: fieldId,
  });

  const result = await DatabaseEventSetGroupByField(payload);

  return result.unwrap();
}

export async function updateGroup(
  viewId: string,
  group: {
    id: string,
    name?: string,
    visible?: boolean,
  },
): Promise<void> {
  const payload = UpdateGroupPB.fromObject({
    view_id: viewId,
    group_id: group.id,
    name: group.name,
    visible: group.visible,
  });

  const result = await DatabaseEventUpdateGroup(payload);

  return result.unwrap();
}

export async function moveGroup(viewId: string, fromGroupId: string, toGroupId: string): Promise<void> {
  const payload = MoveGroupPayloadPB.fromObject({
    view_id: viewId,
    from_group_id: fromGroupId,
    to_group_id: toGroupId,
  });

  const result = await DatabaseEventMoveGroup(payload);

  return result.unwrap();
}
